### PR TITLE
support a new MCU:FS026. add a new keyboard:essemi/fs026.

### PR DESCRIPTION
**Pull request by liuhy-2020**
support a new MCU:FS026.
add a new keyboard:essemi/fs026.

Need to update ChibiOS-Contrib(Merged PR:https://github.com/ChibiOS/ChibiOS-Contrib/pull/397)

## Types of Changes

- [v] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [v] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [v] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [ ] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [v] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [v] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by tzarc** [comment]
Half of my review comments on the previous PR still apply.
https://github.com/qmk/qmk_firmware/pull/23233#pullrequestreview-1930548551

**Comment by tzarc** [comment]
Also, why are things ES32 as well as FS026? Should the MCU actually be called ES32FS026, much like [this](https://www.essemi.com/index/article/detail?id=1243)?

**Comment by liuhy-2020** [comment]
The MCU library I am using comes from ES32_SDK.
In ES32_SDK, the MCU is referred to as ES32F0283_FS026.
ES32F0283 and FS026 are different MCUs. I can find them all in the chip package. The majority of peripherals used in ES32F0283 and FS026 are the same. But the resources are different.

**Comment by liuhy-2020** [comment]
The reason for QMK CI Build failure is that ChibiOS Contrib has not been updated to the latest version.

next step： add a corresponding test board under keyboards/handwired/onekey?

**Comment by liuhy-2020** [comment]
@tzarc     Is it convenient to update ChibiOS Contrib? (FS026 merged into ChibiOS Contrib last week)

**Comment by tzarc** [comment]
CI disabled until we update ChibiOS-Contrib.

**Comment by liuhy-2020** [comment]
> CI disabled until we update ChibiOS-Contrib -- #23405.

When will the ChibiOS-Contrib be updated approximately?

**Comment by tzarc** [comment]
> When will the ChibiOS-Contrib be updated approximately?

Already done.

File: data/schemas/keyboard.jsonschema [hunk 1]
```diff
@@ -90,6 +90,7 @@
                 "STM32L442",
                 "STM32L443",
                 "GD32VF103",
+                "FS026",
```
Comment (tzarc) [inline data/schemas/keyboard.jsonschema:93]: Remove this change. There's only one board with this at the moment, and we'll determine if/when we need to add an explicit entry. Use https://github.com/qmk/qmk_firmware/blob/master/keyboards/kinesis/kint41/rules.mk for an example on configuring a board without having corresponding entries in this file.

File: keyboards/essemi/fs026/config.h [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#define MATRIX_UNSELECT_DRIVE_HIGH
+
+#define DYNAMIC_KEYMAP_EEPROM_MAX_ADDR 1023
+#define TRANSIENT_EEPROM_SIZE 1024
```
Comment (tzarc) [inline keyboards/essemi/fs026/config.h:22]: Remove these.

File: keyboards/essemi/fs026/fs026.c [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+                             {1, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 1, 1, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 1, 1, 1, 1, 1, 4, 1, 1, 1, 1, 1, 1}};
```
Comment (tzarc) [inline keyboards/essemi/fs026/fs026.c:17]: This should be in `keyboard.json`.

File: keyboards/essemi/fs026/halconf.h [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+#define HAL_USE_PWM TRUE
+
+#define ES32_PWM_USE_GP16C4T2 TRUE
+
+#define BACKLIGHT_PWM_DRIVER PWM_GP16C4T2
+#define BACKLIGHT_PWM_CHANNEL 3
+
+#define HAL_USE_USB TRUE
+#define HAL_USE_PAL TRUE
+
```
Comment (tzarc) [inline keyboards/essemi/fs026/halconf.h:28]: Unnecessary.

File: keyboards/essemi/fs026/halconf.h [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+#pragma once
+
+#define HAL_USE_PWM TRUE
+
+#define ES32_PWM_USE_GP16C4T2 TRUE
+
+#define BACKLIGHT_PWM_DRIVER PWM_GP16C4T2
+#define BACKLIGHT_PWM_CHANNEL 3
+
```
Comment (tzarc) [inline keyboards/essemi/fs026/halconf.h:25]: These should be in `config.h`

File: keyboards/essemi/fs026/mcuconf.h
Comment (tzarc) [inline keyboards/essemi/fs026/mcuconf.h:1]: Delete this file.

File: keyboards/essemi/fs026/readme.md [hunk 6]
```diff
@@ -0,0 +1,24 @@
+# FS026
+
+Supported Hardware: *FS026 PDS*.
+
+This is a 17x6 keyboard with WS2812B key lighting.
+It's designed for a modular keyboard mount using an FSO26.
```
Comment (tzarc) [inline keyboards/essemi/fs026/readme.md:6]: Probably want to make sure you get this correct, given you're attempting to introduce support for it.
```suggestion
It's designed for a modular keyboard mount using an FS026.
```

File: keyboards/essemi/fs026/readme.md [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+* Hardware Supported: *ES-PDS, FS026*
+* Hardware Availability: *https://www.essemi.com/*
+
+Make example for this keyboard (after setting up your build environment):
+
+    make essemi\fs026:default
+
+Flashing example:
+
+    make essemi\fs026:default:flash(msc device or qmk_toolbox)
```
Comment (tzarc) [inline keyboards/essemi/fs026/readme.md:18]: Wrong slashes.
Flashing example is wrong format, use the standard template.

File: keyboards/essemi/fs026/readme.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+## Bootloader
+
+Enter the bootloader in 1 ways:
+
+* **Bootmagic reset**: Hold down the key at Esc in the matrix and plug in the keyboard.
```
Comment (tzarc) [inline keyboards/essemi/fs026/readme.md:24]: This is incorrect -- the keyboard is marked with a custom bootloader but has no bootloader entry implementation.

File: keyboards/essemi/fs026/rules.mk [hunk 9]
```diff
@@ -0,0 +1,8 @@
+# Enter lower-power sleep mode when on the ChibiOS idle thread
+OPT_DEFS += -DCORTEX_ENABLE_WFI_IDLE=FALSE
+
```
Comment (tzarc) [inline keyboards/essemi/fs026/rules.mk:3]: Delete these.

File: keyboards/essemi/fs026/rules.mk [hunk 10]
```diff
@@ -0,0 +1,8 @@
...
+# Board: it should exist either in <chibios>/os/hal/boards/
+#  or <this_dir>/boards
+BOARD = FS026
+
```
Comment (tzarc) [inline keyboards/essemi/fs026/rules.mk:7]: Move the configuration from `mcu_selection.mk` to here.

File: lib/python/qmk/constants.py [hunk 11]
```diff
@@ -22,7 +22,7 @@
 MAX_KEYBOARD_SUBFOLDERS = 5
 
 # Supported processor types
-CHIBIOS_PROCESSORS = 'cortex-m0', 'cortex-m0plus', 'cortex-m3', 'cortex-m4', 'MKL26Z64', 'MK20DX128', 'MK20DX256', 'MK64FX512', 'MK66FX1M0', 'RP2040', 'STM32F042', 'STM32F072', 'STM32F103', 'STM32F303', 'STM32F401', 'STM32F405', 'STM32F407', 'STM32F411', 'STM32F446', 'STM32G431', 'STM32G474', 'STM32H723', 'STM32H733', 'STM32L412', 'STM32L422', 'STM32L432', 'STM32L433', 'STM32L442', 'STM32L443', 'GD32VF103', 'WB32F3G71', 'WB32FQ95'
+CHIBIOS_PROCESSORS = 'cortex-m0', 'cortex-m0plus', 'cortex-m3', 'cortex-m4', 'MKL26Z64', 'MK20DX128', 'MK20DX256', 'MK64FX512', 'MK66FX1M0', 'RP2040', 'STM32F042', 'STM32F072', 'STM32F103', 'STM32F303', 'STM32F401', 'STM32F405', 'STM32F407', 'STM32F411', 'STM32F446', 'STM32G431', 'STM32G474', 'STM32H723', 'STM32H733', 'STM32L412', 'STM32L422', 'STM32L432', 'STM32L433', 'STM32L442', 'STM32L443', 'GD32VF103', 'WB32F3G71', 'WB32FQ95', 'FS026'
```
Comment (tzarc) [inline lib/python/qmk/constants.py:25]: Revert this change.

File: platforms/chibios/boards/FS026/configs/board.h
Comment (tzarc) [inline platforms/chibios/boards/FS026/configs/board.h:1]: Delete this file.

File: platforms/chibios/mcu_selection.mk
Comment (tzarc) [inline platforms/chibios/mcu_selection.mk:1]: Revert the changes to this file, move them to keyboard-level `rules.mk`.

File: keyboards/essemi/fs026/info.json
Comment (tzarc) [inline keyboards/essemi/fs026/info.json:1]: This should now be named `keyboard.json` not `info.json`.

File: platforms/chibios/boards/FS026/configs/mcuconf.h
Comment (tzarc) [inline platforms/chibios/boards/FS026/configs/mcuconf.h:1]: Move this file to be inside the keyboard directory.

File: platforms/chibios/boards/FS026/configs/config.h
Comment (tzarc) [inline platforms/chibios/boards/FS026/configs/config.h:1]: Delete this file.

File: platforms/chibios/boards/FS026/board/board.mk
Comment (tzarc) [inline platforms/chibios/boards/FS026/board/board.mk:1]: Delete this file.

File: keyboards/essemi/fs026/rules.mk [hunk 12]
```diff
@@ -0,0 +1,8 @@
...
+#EEPROM_DRIVER = custom
```
Comment (tzarc) [inline keyboards/essemi/fs026/rules.mk:8]: Unnecessary.

File: keyboards/essemi/fs026/info.json [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+      "static_gradient": true,
+      "rgb_test": true,
+      "alternating": true,
+      "twinkle": true
+    }
+  },
+  "ws2812": {
+    "pin": "C0"
+  },
+    "processor": "FS026",
```
Comment (tzarc) [inline keyboards/essemi/fs026/info.json:48]: Unnecessary as this is specified in `rules.mk`.
```suggestion
```

File: keyboards/essemi/fs026/chconf.h [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
```
Comment (tzarc) [inline keyboards/essemi/fs026/chconf.h:16]:
```suggestion
// Copyright 2024 Shanghai Eastsoft Microelectronics Co., Ltd
// SPDX-License-Identifier: GPL-2.0-or-later
```

File: keyboards/essemi/fs026/config.h [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
```
Comment (tzarc) [inline keyboards/essemi/fs026/config.h:16]:
```suggestion
// Copyright 2024 Shanghai Eastsoft Microelectronics Co., Ltd
// SPDX-License-Identifier: GPL-2.0-or-later
```

File: keyboards/essemi/fs026/fs026.c [hunk 16]
```diff
@@ -0,0 +1,19 @@
+
```
Comment (tzarc) [inline keyboards/essemi/fs026/fs026.c:1]:
```suggestion
// Copyright 2024 Shanghai Eastsoft Microelectronics Co., Ltd
// SPDX-License-Identifier: GPL-2.0-or-later
```

File: keyboards/essemi/fs026/halconf.h [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
```
Comment (tzarc) [inline keyboards/essemi/fs026/halconf.h:16]:
```suggestion
// Copyright 2024 Shanghai Eastsoft Microelectronics Co., Ltd
// SPDX-License-Identifier: GPL-2.0-or-later
```

File: keyboards/essemi/fs026/keymaps/default/keymap.c [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
```
Comment (tzarc) [inline keyboards/essemi/fs026/keymaps/default/keymap.c:15]:
```suggestion
// Copyright 2024 Shanghai Eastsoft Microelectronics Co., Ltd
// SPDX-License-Identifier: GPL-2.0-or-later
```

**Review by tzarc — CHANGES_REQUESTED** [review]
After everything has been applied, the only changes in this PR should be under `keyboards/essemi`. No other files need to be modified at this stage.

File: keyboards/essemi/fs026/fs026.c [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+            /*Remap EFLASH,remap to boot*/
+            md_fc_lock();
+            md_syscfg_set_memory_mapping(SYSCFG, MD_SYSCFG_MEMMOD_MAIN);
+            md_syscfg_set_flash_remap_base(SYSCFG, 0);
+            md_syscfg_enable_memory_remap(SYSCFG);
+
+            NVIC_SystemReset();
+        }
+    }
+}
```
Comment (tzarc) [inline keyboards/essemi/fs026/fs026.c:57]: Checking matrix `(0,0)` is handled elsewhere in QMK. Assuming your HAL in ChibiOS-Contrib is correct, you just need to provide a bootloader jump.

```suggestion
void bootloader_jump(void) {
    /*Remap EFLASH,remap to boot*/
    md_fc_lock();
    md_syscfg_set_memory_mapping(SYSCFG, MD_SYSCFG_MEMMOD_MAIN);
    md_syscfg_set_flash_remap_base(SYSCFG, 0);
    md_syscfg_enable_memory_remap(SYSCFG);
    NVIC_SystemReset();
}

void mcu_reset(void) {
    NVIC_SystemReset();
}
```

File: keyboards/essemi/fs026/rules.mk [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+  # Linker script to use
+  # - it should exist either in <chibios>/os/common/startup/ARMCMx/compilers/GCC/ld/
+  #   or <keyboard_dir>/ld/
+  MCU_LDSCRIPT ?= FS026
+
+  # Startup code to use
+  #  - it should exist in <chibios>/os/common/startup/ARMCMx/compilers/GCC/mk/
+  MCU_STARTUP ?= FS026
+
+  USE_FPU ?= no
```
Comment (tzarc) [inline keyboards/essemi/fs026/rules.mk:29]: Please remove the indentation.

File: keyboards/essemi/fs026/halconf.h [hunk 21]
```diff
@@ -0,0 +1,10 @@
+// Copyright 2024 Shanghai Eastsoft Microelectronics Co., Ltd
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#define HAL_USE_PWM TRUE
+
+#define ES32_PWM_USE_GP16C4T2 TRUE
```
Comment (tzarc) [inline keyboards/essemi/fs026/halconf.h:8]: This should be in `mcuconf.h`.

File: keyboards/essemi/fs026/keyboard.json [hunk 22]
```diff
@@ -0,0 +1,154 @@
+{
+    "manufacturer": "essemi",
```
Comment (tzarc) [inline keyboards/essemi/fs026/keyboard.json:2]:
```suggestion
    "manufacturer": "essemi",
    "maintainer": "liuhy-2020",
```

File: keyboards/essemi/fs026/mcuconf.h [hunk 23]
```diff
@@ -0,0 +1,162 @@
...
+#define ES32_ICU_TIM4_IRQ_PRIORITY 7
+
+/*
+ * PWM driver system settings.
+ */
+#define ES32_PWM_USE_ADVANCED FALSE
+#define ES32_PWM_USE_TIM1 FALSE
+#define ES32_PWM_USE_TIM2 FALSE
+#define ES32_PWM_USE_TIM3 FALSE
+#define ES32_PWM_USE_TIM4 FALSE
```
Comment (tzarc) [inline keyboards/essemi/fs026/mcuconf.h:95]: These are written as TIM1...TIM4, but the backlight is called GP16C4T2.
The defines don't match https://github.com/ChibiOS/ChibiOS-Contrib/blob/chibios-21.11.x/os/hal/ports/ES32/LLD/TIMv1/hal_pwm_lld.h

Please fix up this file so that ALL item names are actually correct, not just the PWM ones.

File: keyboards/essemi/fs026/keyboard.json
Comment (tzarc) [inline keyboards/essemi/fs026/keyboard.json:1]: When all other changes to this file are made, run `qmk format-json -i` on this file.

File: keyboards/essemi/fs026/keymaps/default/keymap.c
Comment (tzarc) [inline keyboards/essemi/fs026/keymaps/default/keymap.c:1]: Please arrange these layers so that they roughly resemble a TKL.
See https://github.com/qmk/qmk_firmware/blob/cd0043de8ffaa1cdbeb86f242b250c8d23f5a231/layouts/default/tkl_ansi/default_tkl_ansi/keymap.c#L23-L31 for an example.

**Review by tzarc — CHANGES_REQUESTED** [review]
More changes.

File: keyboards/essemi/fs026/fs026.c (hunk 19 above)
Comment (tzarc) [reply keyboards/essemi/fs026/fs026.c:57]: If you need to use 2-stage bootloader in order to deal with system reset, use https://github.com/qmk/qmk_firmware/blob/master/platforms/chibios/bootloaders/wb32_dfu.c as an example instead.

**Comment by liuhy-2020** [comment]
What modifications are needed?

**Comment by tzarc** [comment]
> What modifications are needed?

Your `mcuconf.h` is now in the wrong format and only configures PWM entries, but should be configuring every available peripheral. You also have not addressed the issue with the readme.

**Comment by liuhy-2020** [comment]
If the MCU driver currently only supports GPIO, USB, and PWM functions. Is this okay?

In the future, improve MCU drivers.

**Comment by tzarc** [comment]
This is what a normal `mcuconf.h` would look like for a typical Chinese STM32 clone: https://github.com/qmk/qmk_firmware/blob/master/platforms/chibios/boards/GENERIC_WB32_F3G71XX/configs/mcuconf.h

Note that it includes all peripheral types. This is the expectation, even if they're not yet implemented in ChibiOS-Contrib.

I understand you've only implemented the bare minimum in the base layer, but we don't want to be modifying and adapting these files every time you onboard a new peripheral.

As it stands the only peripheral you _have_ added is PWM, so technically the default should be that USB and GPIO are disabled and your keyboard shouldn't work. As such, we'll assume you've misconfigured your code in ChibiOS-Contrib if they're working without being configured. Might be best fixing that up -- peripherals should be disabled by default in the base layer.

**Comment by liuhy-2020** [comment]
OK.I will add all peripherals in mcuconf. h.

**Comment by liuhy-2020** [comment]
@tzarc  What other files need to be modified？

**Comment by liuhy-2020** [comment]
@tzarc  PR is still in review?

**Comment by liuhy-2020** [comment]
@tzarc

**Comment by zvecr** [comment]
Please stop pinging tzarc, and take note of the message we use on Discord.

> While we know that your PR is important to you, it may take some time for QMK collaborators to get to it.
>
> PR reviews are something that is done in their free time -- collaborators are not paid nor compensated for the time spent reviewing, as it is a labor of love. As such, this means that it can take time for them to get to your PR -- things like family, or life can get in the way of finding time to perform reviews, and burnout is a serious consideration. The QMK firmware repository averages 300 PRs opened and 300 PRs merged every month, so please have patience.

**Comment by liuhy-2020** [comment]
ok
